### ISSUE TITLE
Pruning & Pytorch (Possible Bug)

### ISSUE THREAD

**Issue by Serendipity31** [optuna/optuna#5068]
### Expected behavior

I accidentally posted htis in optuna/examples. I think it probably should be here, but I can't delete it, so I'll link to the full description of the issue here: https://github.com/optuna/optuna-examples/issues/219

The tldr is that I cannot get optuna to prune trials in the context of a model that uses the pytorch lightning trainer. 

It may be I've misunderstood the example instructions for how to do this, but it may also be that this integration isn't fully working in light of recent updates. If someone could look into this and provide some guidance, I would really appreciate it.

### Environment

    Optuna version: 3.4.0
    Python version: 3.11.5
    OS: Windows 10 Enterprise
    gluonTS version: 0.13.7
    lighning version: 2.1.0
    pytorch-lightning version: 2.1.0
    torch version: 2.1.0

### Error messages, stack traces, or logs

```shell
None - just no trials are pruned.
```

### Steps to reproduce

Please see: https://github.com/optuna/optuna-examples/issues/219

**Comment by nzw0301** [comment]
Could you share the minimal reproducible code with us for readability?

**Comment by contramundum53** [comment]
@nzw0301 I think that's in https://github.com/optuna/optuna-examples/issues/219

**Comment by nzw0301** [comment]
Thanks, but it is not minimal and reproducible (the script is only objective func).

**Comment by HideakiImamura** [comment]
+1 to @nzw0301 

@Serendipity31 Could you provide the self-contained reproducible example for this?

**Comment by Serendipity31** [comment]
@HideakiImamura and @nzw0301 - yes - sure! The parts I didn't include basically come straight from the [gluonTS example](https://ts.gluon.ai/stable/tutorials/advanced_topics/hp_tuning_with_optuna.html), so I didn't think to include them in my original post. Apologies! Let me know if you need anything else.

Just as a reminder, optuna appears to execute fully with the code as outlined below. It completes trials and returns a best_trial along with associated hyperparameter values. It just never seems to prune any trials, so things take longer than I expect they would if it was pruning trials. 

## GluonTS Tutorial Overview
In the gluonTS example, the 'thing' given to the 'func' parameter within optimize() is an instance of a class that contains an init function, a 'get_params()' function, and then a call function. 

- The init function takes a dataset and uses the gluonTS split() function and a locally defined 'dataentry_to_dataframe()' function to create the training and validation versions of the given dataset. 

- The 'get_params()' function is where the optuna hyperparameter ranges are defined and calls to trial.suggest() are made. 

- The call function is where the model is defined, where the model is trained, and where its performance is evaluated. The model definition happens through creating an instance of the DeepAREstimator. Within the DeepAREstimator, information is passed to the Pytorch Lightning trainer in the 'trainer_kwargs' dictionary argument. A predictor is created by calling train() on the estimator. Forecasts are generated from that predictor. The accuracy of the model is then evaluated using an evaluator function that also comes from gluonTS. 

## Code
I am using an interactive IDE, so I have the code split into different chunks. They get executed in this order. 

Firstly, I import these packages:
```{python}
# Packages required by gluonTS in demo related to Optuna 
import torch
import optuna
from gluonts.dataset.split import split
from gluonts.evaluation import Evaluator
from gluonts.torch.model.deepar import DeepAREstimator

# Additional packages I'm using
import time
import pickle
import matplotlib.pyplot as plt
from pyprojroot.here import here
from lightning.pytorch import seed_everything
from gluonts.dataset.pandas import PandasDataset

# set seeds for numpy, torch and python.random
seed_everything(42, workers=True)
```

Secondly, to source the 'dataentry_to_dataframe()' function, execute this code:
```{python}
def dataentry_to_dataframe(entry):
  df = pd.DataFrame(
     data = entry["target"], 
     columns = [entry.get("item_id")],
     index = pd.period_range(
        start = entry["start"],
        periods = len(entry["target"]),
        freq = entry["start"].freq
      ),
  )
  return df
```
Thirdly, to define the class, I execute the code below (either leaving 'pruning_callback' commented out, or including it, depending on whether one is trying approach 1 or approach 2, respectively). It may be helpful for me to explain a few things:

1. In my case 'dataset' refers to a [gluonTS Pandas DataSet](https://ts.gluon.ai/stable/api/gluonts/gluonts.dataset.pandas.html) that I have saved as a pickle. If someone used a different data structure they would provide values to the exogenous variables in a different way. I can't give you my data, but public example datasets are available through gluonTS that could be used.
2. The gluonTS example doesn't address pruning, but I tried to follow the example shown [here](https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_lightning_simple.py) where the integration with Pytorch Lightning is passed through a callback to Trainer.

```{python}
class DeepARTuningObjective:
  
  def __init__(
    self, dataset, prediction_length=1, metric_type="mean_wQuantileLoss"
  ):
      self.dataset = dataset
      self.freq = dataset.freq
      self.metric_type = metric_type
      self.prediction_length = prediction_length
      
      self.train, test_template = split(dataset, offset=-self.prediction_length)
      
      validation = test_template.generate_instances(prediction_length=self.prediction_length)
      
      self.validation_input = [entry[0] for entry in validation]
      
      self.validation_label = [dataentry_to_dataframe(entry[1]) for entry in validation]

  def get_params(self, trial) -> dict:
    return {
       "num_layers": trial.suggest_int("num_layers", 1, 5),
       "hidden_size": trial.suggest_int("hidden_size", 30, 100),
       "learning_rate": trial.suggest_float("learning_rate", 1e-5, 1e-1, log=True), 
       "weight_decay": trial.suggest_float("weight_decay", 1e-10, 1e-5),
       "dropout_rate": trial.suggest_float("dropout_rate", 0.1, 0.2),
       "batch_size" trial.suggest_int("batch_size", 12, 72, step=12),
       "max_epochs": trial.suggest_int("epochs", 10, 101)
     }
      
  def __call__(self, trial):
    params = self.get_params(trial)
    #pruning_callback = optuna.integration.PyTorchLightningPruningCallback(trial, "val_loss")
    estimator = DeepAREstimator(
      
      #Basic Parameters
      freq = self.freq,
      prediction_length = self.prediction_length,
      context_length = self.prediction_length,
      
      #Hyperparameters to be tuned
      num_layers = params["num_layers"],    
      hidden_size = params["hidden_size"],   
      lr = params["learning_rate"], 
      weight_decay = params["weight_decay"],  
      dropout_rate = params["dropout_rate"],
      batch_size = params["batch_size"],   
      
      #Exogenous Variables
      num_feat_dynamic_real = self.dataset.num_feat_dynamic_real,
      num_feat_static_cat = self.dataset.num_feat_static_cat,
      cardinality = self.dataset.static_cardinalities.tolist(),
      
      #Additional call to pl.Trainer with hyperparameter to be tuned
           trainer_kwargs ={
                "max_epochs": params["max_epochs"]+1,
                "deterministic": True,
                #"callbacks": [pruning_callback]
      },
    )
    
    predictor = estimator.train(self.train, cache_data=True)
    forecast_it = predictor.predict(self.validation_input)
    
    forecasts = list(forecast_it)
    
    evaluator = Evaluator(quantiles=[0.1, 0.5, 0.9])
    agg_metrics, item_metrics = evaluator(
      self.validation_label, forecasts, num_series=len(self.dataset)
    )
    return agg_metrics[self.metric_type] 
```
After executing the code for the appropriate version of the class below, the next code chunk loads the data, calls create_study(), and then calls study.optimize(). There are two slightly different versions of this, each of which is shown below.

### Approach 1: Define 'pruner' and use in optuna.create_study()
In this approach, I define 'pruner', and then used 'pruner' as a parameter in create_study(). However, the version of the DeepARTuningObjective that is used in this case is one without 'pruner_callback'. 
```{python}
pickle_in = open(here("pythonPickles/dataset.pickle"), "rb")
dataset = pickle.load(pickle_in)
pickle_in.close()

pruner = optuna.pruners.MedianPruner()
sampler = optuna.samplers.TPESampler(seed=42)

study = optuna.create_study(
    direction = "minimize",
    pruner  = pruner,
    sampler = sampler,
    study_name = "study_name",
    storage = "sqlite:///optunaStudies/study_name.db",
    load_if_exists = True
)
  
study.optimize(
    func = DeepARTuningObjective(dataset = dataset, prediction_length=1),
    n_trials  = 10
)  
```
### Approach 2: Define 'pruner_callback' in the objective function, and pass it to Trainer as a callback.
In this approach, create_study() is not given a pruner. However, the version of DeepARTuningObjective that is used contains 'pruner_callback'. 
```{python}
pickle_in = open(here("pythonPickles/dataset.pickle"), "rb")
dataset = pickle.load(pickle_in)
pickle_in.close()

sampler = optuna.samplers.TPESampler(seed=42)

study = optuna.create_study(
    direction = "minimize",
    sampler = sampler,
    study_name = "study_name",
    storage = "sqlite:///optunaStudies/study_name.db",
    load_if_exists = True
)
  
study.optimize(
    func = DeepARTuningObjective(dataset = dataset, prediction_length=1),
    n_trials  = 10
)  
```
### Approach 3: Try both simultaneously
I also tried approach 1 and 2 simultaneously (using 'pruner' and 'pruner_callback'). This neither produced an error nor pruning.

**Comment by HideakiImamura** [comment]
Thanks for the detailed explanation. Let me add the contribution welcome issue label.

**Comment by nzw0301** [comment]
@Serendipity31 I'm afraid that the code is still not executable (the code has some invalid syntax and missing import statement) and the pruning line, the most important part `optuna.integration.PyTorchLightningPruningCallback`, is commented out, so I'm not sure how to reproduce it to resolve this issue. 

Please share the executable code snippet instead of separated code blocks like above, which makes us hard to investigate the issue.

**Comment by Serendipity31** [comment]
@nzw0301 - I'm not sure what you mean. I'm using an IDE that runs python in an interactive way. I provided the separate code chunks that I have above. Each of them executes individually and sequentially. They also don't throw any errors on my end (just they don't deliver pruning), so there shouldn't be any syntax problems unless my copy/paste didn't fully catch one of the lines. Can you please be more specific?

Regarding your other point, I explained above why those two lines in the definintion of `DeepARTuningObjective` are commented out. I tried two ways of getting pruning to work. One follows an example that created a `pruner` and used it as an argument in `optuna.create_study` (Approach 1). The other follows an example that passes a callback to Pytorch Lightning `Trainer` (Approach 2). 

To try approach 1, leave the lines in the definition of `DeepARTuningObjective` commented out when intitialising that class. Then execute the code snippet under 'Approach 1' header. 

To try approach 2, uncomment those lines in the definition of `DeepARTuningObjective` and then initialise that class. Then execute the code snippet under the 'Approach 2' header. 

To try both approaches simultaneously, uncomment those lines in the definition of `DeepARTuningObjective` and then initialise that class. Then execute the code snippet under the 'Approach 1' header.

**Comment by nzw0301** [comment]
The third block is not executable due to the bug and the second block is using `pd` that is not imported.

**Comment by Serendipity31** [comment]
@nzw0301 - Thanks for pointing that out. It seems I imported pandas in an earlier code chunk. I'm not sure how that ':' was mising. I've edited the above the reflect this. 

You may also need to create a .here file in your working directory for the here() function from pyprojroot.here to work.

**Comment by nzw0301** [comment]
Thank you for updating the code blocks. Likewise, such manual procedures (concretely, `here` and pickle data creation) make it hard to run the code you share. Could you provide a standalone Python script? Here, I mean the code is executable pickle data creation should be in the code too or using the dataset, gluonTS Pandas DataSet, directly, not via loading locally save pickled file).

I think approach 1 doesn't perform pruning because the objective does not call `report` and `prune`. The approaches 2 and 3 are identical because if we don't specify pruner at `create_study`, `MedianPruner` is used. Thus, approach 2 or approach 3's code is enough to be investigated.

**Comment by Serendipity31** [comment]
@nzw0301 - Sorry this is proving to be difficult, and thanks for your patience. I'm not at all used to creating standalone Python scripts. The code block below uses a gluonTS public dataset, and is basically a copy/paste from the [gluonTS tutorial](https://ts.gluon.ai/stable/tutorials/advanced_topics/hp_tuning_with_optuna.html), except for the parts related to pruning.

When I updated gluonTS yesterday to version 0.14.0, I started getting an error when trying to use the callback as shown above `ValueError("Expected a parent")`. I found [this discussion](https://github.com/optuna/optuna/issues/4689) that provides a kind of workaround for this issue. You'll see that is reflected in the code below. 

I'm not sure how one could call `report` in this. The callback doesn't seem to have the option of taking that as an input, and the training step isn't something that I define myself. 

The training step is defined [here](https://github.com/awslabs/gluonts/blob/19a4893233099c031c3eb868c56f3140d9b5cdb1/src/gluonts/torch/model/deepar/lightning_module.py#L70). I've noticed that the training step returns `train_loss`, but doesn't return a dictionary. The Optuna documentation on the [PyTorch Lightning integration](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.PyTorchLightningPruningCallback.html) seems to suggest it's expecting a dictionary entry that can be retrieved via the key. Could that be (part of) the issue?

**Package Versions**
torch: 2.1.0
gluonts: 0.14.0
numpy: 1.26.1
optuna: 3.4.0
pandas: 2.1.1
lighnting: 2.1.0
pytoch-lightning: 2.1.0

**Code**
```{python}
import time
import optuna
import torch
import gluonts
import numpy as np
import pandas as pd
import lightning.pytorch as pl
from gluonts.dataset.split import split
from gluonts.evaluation import Evaluator
from gluonts.dataset.repository import get_dataset
from optuna.integration import PyTorchLightningPruningCallback
from optuna.pruners import MedianPruner
################################################################################

dataset = get_dataset("m4_hourly")

################################################################################

def dataentry_to_dataframe(entry):
    df = pd.DataFrame(
        entry["target"],
        columns=[entry.get("item_id")],
        index=pd.period_range(
            start=entry["start"], periods=len(entry["target"]), freq=entry["start"].freq
        ),
    )

    return df

################################################################################
# Thread discussing need for custom class to get Optuna and Pytorch talking for 
# pruning: https://github.com/optuna/optuna/issues/4689

class OptunaPruning(PyTorchLightningPruningCallback, pl.Callback):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

################################################################################
class DeepARTuningObjective:
    def __init__(
        self, dataset, prediction_length, freq, metric_type="mean_wQuantileLoss"
    ):
        self.dataset = dataset
        self.prediction_length = prediction_length
        self.freq = freq
        self.metric_type = metric_type

        self.train, test_template = split(dataset, offset=-self.prediction_length)
        validation = test_template.generate_instances(
            prediction_length=prediction_length
        )
        self.validation_input = [entry[0] for entry in validation]
        self.validation_label = [
            dataentry_to_dataframe(entry[1]) for entry in validation
        ]

    def get_params(self, trial) -> dict:
        return {
            "num_layers": trial.suggest_int("num_layers", 1, 5),
            "hidden_size": trial.suggest_int("hidden_size", 10, 50),
        }

    def __call__(self, trial):
        params = self.get_params(trial)
        estimator = DeepAREstimator(
            num_layers=params["num_layers"],
            hidden_size=params["hidden_size"],
            prediction_length=self.prediction_length,
            freq=self.freq,
            trainer_kwargs={
                "enable_progress_bar" : False,
                "enable_model_summary": False,
                "max_epochs": 10,
                "callbacks": [OptunaPruning(trial, monitor="train_loss")]
            },
        )

        predictor = estimator.train(self.train, cache_data=True)
        forecast_it = predictor.predict(self.validation_input)

        forecasts = list(forecast_it)

        evaluator = Evaluator(quantiles=[0.1, 0.5, 0.9])
        agg_metrics, item_metrics = evaluator(
            self.validation_label, forecasts, num_series=len(self.dataset)
        )
        return agg_metrics[self.metric_type]

################################################################################
pruner = optuna.pruners.MedianPruner(n_startup_trials=2)

study = optuna.create_study(
  direction  = "minimize",
  study_name = "test-study-pruner",
  pruner     = pruner
)

start_time = time.time()
study.optimize(
    func = DeepARTuningObjective(dataset.train, dataset.metadata.prediction_length, dataset.metadata.freq),
    gc_after_trial = True,
    n_trials = 5,
    timeout = 600,
)
end_time = time.time()
print((end_time - start_time)/60)

################################################################################
pruned = [t for t in study.trials if t.state == optuna.trial.TrialState.PRUNED]
completed = [t for t in study.trials if t.state == optuna.trial.TrialState.COMPLETE]

print("Summary statistics: ")
print(" Number of finished trials: " , len(study.trials))
print(" Number of pruned trials: "   , len(pruned))
print(" Number of completed trials: ", len(completed))

print("Best Trial:")
best_trial = study.best_trial

print(" Accuracy Value: ", best_trial.value)

print(" Params: ")
for key, value in best_trial.params.items():
  print(" {}: {}".format(key, value))
  
```

**Comment by nzw0301** [comment]
Thanks. Could you specify the validation set?

**Comment by nzw0301** [comment]
As in the log message on my env
> /opt/homebrew/Caskroom/miniconda/base/envs/optuna/lib/python3.10/site-packages/lightning/pytorch/trainer/configuration_validator.py:74: You defined a `validation_step` but have no `val_dataloader`. Skipping val loop.

the validation step is skipped, so the pruning never happens.

**Comment by nzw0301** [comment]
Thanks to the clarification, I believe that this is not a bug of optuna. Let me convert this issue into a discussion.
